tooling/nargo_fmt/src/formatter.rs

```diff
@@ -0,0 +1,303 @@
+    ///
+    /// Before writing the token any comments and spaces are skipped. This is so that
+    /// a caller can call `write_token`, `write_keyword`, `write_space`, etc., without
+    /// having to explicitly call `skip_comments_and_whitespace` in between those calls.
+    fn write_token(&mut self, token: Token) {
+        self.skip_comments_and_whitespace();
+        if self.token == token {
+            self.write_current_token();
+            self.bump();
+        } else {
+            panic!("Expected token {:?}, got: {:?}", token, self.token);
+        }
+    }
+
+    /// Writes the current token but doesn't advance to the next one.
+    fn write_current_token(&mut self) {
+        self.write(&self.token.to_string());
+    }
+
+    /// Writes the current token trimming its end but doesn't advance to the next one.
+    /// Mainly used when writing comment lines, because we never want trailing spaces
+    /// inside comments.
+    fn write_current_token_trimming_end(&mut self) {
+        self.write(self.token.to_string().trim_end());
+    }
+
+    /// Writes the current token but without turning it into a string using `to_string()`.
+    /// Instead, we check the token's span and format what's in the original source there
+    /// (useful when formatting integer tokens, because a token like 0xFF ends up being an
+    /// integer with a value 255, but we don't want to change 0xFF to 255).
+    fn write_current_token_as_in_source(&mut self) {
+        self.write_source_span(self.token_span);
+    }
+
+    /// Writes whatever is in the given span relative to the file's source that's being formatted.
+    fn write_source_span(&mut self, span: Span) {
+        self.write(&self.source[span.start() as usize..span.end() as usize]);
+    }
+
+    /// Writes the current indentation to the buffer, but only if the buffer
+    /// does not end with a space (otherwise we'd be indenting too much).
+    fn write_indentation(&mut self) {
+        if self.buffer.ends_with(' ') {
+            return;
+        }
+
+        for _ in 0..self.indentation {
+            for _ in 0..self.config.tab_spaces {
+                self.write(" ");
+            }
+        }
+    }
+
+    /// Writes whatever is in the source at the given span without formatting it,
+    /// then advances the lexer until past the end of the span.
+    /// This is mainly used to avoid formatting items and statements when a
+    /// `noir-fmt:ignore` comment is found.
+    pub(super) fn write_and_skip_span_without_formatting(&mut self, span: Span) {
+        self.write_source_span(span);
+
+        while self.token_span.start() < span.end() {
+            self.bump();
+        }
+    }
+
+    /// Writes a string to the buffer.
+    /// This is the only method that directly appends to the buffer and keeps
+    /// track of the current line width.
+    /// If adding new methods that write to the buffer, always use this method
+    /// instead of directly appending to the buffer.
```

Comment:
Would it make sense to encapsulate the buffer and the current line width in a separate structure to enforce this?

Reply:
Probably yes. I'll try it after addressing the other comments.

Reply:
Done! Thank you, it ended up making the code much clearer.